src/main/scala/org/ergoplatform/nodeView/history/storage/modifierprocessors/HeadersProcessor.scala

```diff
@@ -257,34 +258,77 @@ trait HeadersProcessor extends ToDownloadProcessor with ScorexLogging with Score
+        storeEip37ActivationHeight(parentHeight + 1)
+      }
+    }
+
+    if (parentHeight > 843776 && parentHeight + 1 >= eip37VotedOn.getOrElse(Int.MaxValue)) {
+      // by eip37VotedOn definition could be on mainnet only
+      val epochLength = 128 // epoch length after EIP-37 activation
+      if (parentHeight % epochLength == 0) {
+        val heights = difficultyCalculator.previousHeadersRequiredForRecalculation(parentHeight + 1, epochLength)
+        // todo: if parent is on best chain, read headers directly, not via headerChainBack
+        val chain = headerChainBack(heights.max - heights.min + 1, parent, _ => false)
+        val headers = chain.headers.filter(p => heights.contains(p.height))
+        difficultyCalculator.eip37Calculate(headers, epochLength)
+      } else {
+        parent.requiredDifficulty
+      }
     } else {
-      val parentHeight = parent.height
-
-      if(parentHeight % settings.chainSettings.epochLength == 0) {
-        //todo: it is slow to read thousands headers from database for each header
-        //todo; consider caching here
-        //todo: https://github.com/ergoplatform/ergo/issues/872
-        val heights = difficultyCalculator.previousHeadersRequiredForRecalculation(parentHeight + 1)
-          .ensuring(_.last == parentHeight)
-        if (heights.lengthCompare(1) == 0) {
-          difficultyCalculator.calculate(Array(parent))
+      if (parentHeight == settings.chainSettings.voting.version2ActivationHeight ||
+          parent.height + 1 == settings.chainSettings.voting.version2ActivationHeight) {
+        // Set difficulty for version 2 activation height (where specific difficulty is needed due to PoW change)
+        settings.chainSettings.initialDifficultyVersion2
+      } else {
+        val epochLength = settings.chainSettings.epochLength
+
+        if (parentHeight % epochLength == 0) {
+          //todo: it is slow to read thousands headers from database for each header
+          //todo; consider caching here
+          //todo: https://github.com/ergoplatform/ergo/issues/872
+          val heights = difficultyCalculator.previousHeadersRequiredForRecalculation(parentHeight + 1, epochLength)
+            .ensuring(_.last == parentHeight)
+          if (heights.lengthCompare(1) == 0) {
+            difficultyCalculator.calculate(Array(parent), epochLength)
+          } else {
+            val chain = headerChainBack(heights.max - heights.min + 1, parent, _ => false)
+            val headers = chain.headers.filter(p => heights.contains(p.height))
+            difficultyCalculator.calculate(headers, epochLength)
+          }
         } else {
-          val chain = headerChainBack(heights.max - heights.min + 1, parent, _ => false)
-          val headers = chain.headers.filter(p => heights.contains(p.height))
-          difficultyCalculator.calculate(headers)
+          parent.requiredDifficulty
         }
-      } else {
-        parent.requiredDifficulty
       }
     }
   }
```

Comment:
I really worry about complexity of this method. Can we make old version unchanged and be invoked only before activation? At the same the new logic to be written separately and to be invoked AFTER activation. We need to have some guarantee that before activation the old code works unchanged. And the new logic will be easier to analyze in isolation.
Also complex conditions can be encapsulated into method with descriptive names and ScalaDoc.

Reply:
I made manual testing, syncing from height 600K+ till current moment. Thus I checked how old code works before the activation. To test activation, I used parameter #8, as ~100% blocks are voting for increasing it. Thus I got activation done on my client. 

Please note, both versions will be called after activation, condition `parentHeight > 843776 && parentHeight + 1 >= eip37ActivationHeight.getOrElse(Int.MaxValue)` will be changed in the next version if activation will happen (so we can hardcode activation height).